Caboodle/Permissions/Permissions.android.cs

```diff
@@ -0,0 +1,151 @@
+
+            var context = Platform.CurrentContext;
+
+            foreach (var ap in androidPermissions)
+            {
+                var packageInfo = context.PackageManager.GetPackageInfo(context.PackageName, PackageInfoFlags.Permissions);
+                var requestedPermissions = packageInfo?.RequestedPermissions;
+
+                // If the manifest is missing any of the permissions we need, throw
+                if (!requestedPermissions?.Any(r => r.Equals(ap, StringComparison.OrdinalIgnoreCase)) ?? false)
+                    throw new PermissionException($"You need to declare the permission: `{ap}` in your AndroidManifest.xml");
+            }
+        }
+
+        static Task<PermissionStatus> PlatformCheckStatusAsync(PermissionType permission)
+        {
+            PlatformEnsureDeclared(permission);
+
+            // If there are no android permissions for the given permission type
+            // just return granted since we have none to ask for
+            var androidPermissions = permission.ToAndroidPermissions();
+            if (androidPermissions == null || !androidPermissions.Any())
+                return Task.FromResult(PermissionStatus.Granted);
+
+            var context = Platform.CurrentContext;
+            var targetsMOrHigher = context.ApplicationInfo.TargetSdkVersion >= Android.OS.BuildVersionCodes.M;
+
+            foreach (var ap in androidPermissions)
+            {
+                if (targetsMOrHigher)
+                {
+                    if (context.CheckSelfPermission(ap) != Permission.Granted)
+                        return Task.FromResult(PermissionStatus.Denied);
+                }
+                else
+                {
+                    if (PermissionChecker.CheckSelfPermission(context, ap) != PermissionChecker.PermissionGranted)
+                        return Task.FromResult(PermissionStatus.Denied);
+                }
+            }
+
+            return Task.FromResult(PermissionStatus.Granted);
+        }
+
+        static async Task<PermissionStatus> PlatformRequestAsync(PermissionType permission)
+        {
+            // Check status before requesting first
+            if (await PlatformCheckStatusAsync(permission) == PermissionStatus.Granted)
+                return PermissionStatus.Granted;
+
+            TaskCompletionSource<PermissionStatus> tcs;
+            var doRequest = true;
+            var requestCode = 0;
+
+            lock (locker)
+            {
+                if (requests.ContainsKey(permission))
+                {
+                    tcs = requests[permission].tcs;
+                    doRequest = false;
+                }
+                else
+                {
+                    tcs = new TaskCompletionSource<PermissionStatus>();
+
+                    // Get new request code and wrap it around for next use if it's going to reach max
+                    if (++requestCode >= int.MaxValue)
+                        requestCode = 1;
```

Comment:
What is the scope of `requestCode`? It appears to be created in this method, so it will always be `0` - see line 79.

Reply:
🤦‍♂️